### PR TITLE
module/tectonic: Add default storage classes

### LINKED ISSUE #711: platforms/*: install the default storage classes
Since we don't run the add-on manager we don't' install the default storage classes. We should do this!

- [AWS](https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/storage-class/aws/default.yaml)
- [OpenStack](https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/storage-class/openstack/default.yaml)
- [etc](https://github.com/kubernetes/kubernetes/tree/master/cluster/addons/storage-class)

cc @polvi

### PR DESCRIPTION
**Pull request by kalmog**
Fixes #INST-284
Fixes #711

### CODE REVIEW

**Comment by kalmog** [comment]
test this please

**Comment by kalmog** [comment]
test this please

File: Documentation/variables/aws.md [hunk 1]
```diff
@@ -6,6 +6,10 @@ This document gives an overview of variables used in the AWS platform of the Tec
 
 | Name | Description | Type | Default |
 |------|-------------|:----:|:-----:|
+| extra_storage_params | Set to true, if extra parmeters are needed for the StorageClass definition | string | `true` |
+| extra_storage_params_list | A list of extra storage parameters to add to the storage class configuration (platform dependent), formatted as a list of yaml values.<br><br>Example: [ "type: gp2", "diskformat: thin" ] | list | `<list>` |
+| storage_class_provisioner | The provisioner to use for creating the storage class | string | `kubernetes.io/aws-ebs` |
+| storage_metadata_name | A metadata name tag for the storage class | string | `gp2` |
```
Comment (s-urbaniak) [inline Documentation/variables/aws.md]: the convention is to prefix all top-level variables with `tectonic_`. let's rename `extra_storage_params` to `tectonic_extra_storage_params`, etc.

File: Documentation/variables/aws.md [hunk 2]
```diff
@@ -6,6 +6,10 @@ This document gives an overview of variables used in the AWS platform of the Tec
 
 | Name | Description | Type | Default |
 |------|-------------|:----:|:-----:|
+| extra_storage_params | Set to true, if extra parmeters are needed for the StorageClass definition | string | `true` |
+| extra_storage_params_list | A list of extra storage parameters to add to the storage class configuration (platform dependent), formatted as a list of yaml values.<br><br>Example: [ "type: gp2", "diskformat: thin" ] | list | `<list>` |
```
Comment (s-urbaniak) [inline Documentation/variables/aws.md]: naming question: is it necessary to prefix with `extra_`? or would `tectonic_storage_params` be more concise?

File: modules/tectonic/assets.tf [hunk 3]
```diff
@@ -12,37 +12,45 @@ resource "template_dir" "tectonic" {
...
+    prometheus_operator_base_image      = "${var.container_base_images["prometheus_operator"]}"
+    tectonic_monitoring_auth_base_image = "${var.container_base_images["tectonic_monitoring_auth"]}"
+
+    storage_class_provisioner = "${var.storage_class_provisioner}"
+    storage_metadata_name     = "${var.storage_metadata_name}"
 
+    // We need this hack to generate a conditional yaml block
+    // for storage classes, since the extra parameters differ
+    // from platform to platform
+    storage_parameters = "${var.extra_storage_params ? "parameters:\n  ${indent(2, join("\n", var.extra_storage_params_list))}" : ""}"
```
Comment (enxebre) [inline modules/tectonic/assets.tf]: couldn't we just check if extra_storage_params_list is not empty so no need to coordinate both variables?

File: platforms/gcp/variables.tf [hunk 4]
```diff
@@ -92,3 +92,32 @@ variable "tectonic_gcp_ssh_key" {
   type        = "string"
   description = "(required) Path to an SSH public key file to be provisioned as the SSH key for the 'core' user."
 }
+
+variable "storage_class_provisioner" {
+  description = "The provisioner to use for creating the storage class"
+  default     = "kubernetes.io/gce-pd"
```
Comment (enxebre) [inline platforms/gcp/variables.tf]: just so we are aware this most likely won't work as gcp is currently not using --cloud-provider flag

File: modules/tectonic/assets.tf (hunk 3 above)
Comment (kalmog) [reply modules/tectonic/assets.tf]: a list type parameter cannot be parsed as bool, hence the dual parameters:
```
* data.template_file.init: 1 error(s) occurred:

* data.template_file.init: At column 3, line 1: condition must be type bool, not type list in:

${var.extra_params_list ? "parameters:\n  ${indent(2, join("\n", var.extra_params_list))}" : ""}
```

File: platforms/gcp/variables.tf (hunk 4 above)
Comment (kalmog) [reply platforms/gcp/variables.tf]: Good catch. Similar to #51. I believe we should open a separate ticket for this.

File: modules/tectonic/assets.tf (hunk 3 above)
Comment (squat) [reply modules/tectonic/assets.tf]: @kalmog I believe the suggestions is to use https://www.terraform.io/docs/configuration/interpolation.html#length-list- to determine if the list is empty or not and make decisions based on that

Comment (enxebre) [reply modules/tectonic/assets.tf]: yep something like ```"${length(var.extra_storage_params) > 0 ? "parameters:\n  ${indent(2, join("\n", var.extra_storage_params_list))}" : ""}"```

**Comment by cpanato** [comment]
@squat  @kalmog  this is still valid? if yes, can you please rebase?

**Comment by kalmog** [comment]
@cpanato the logic is likely going to be moved to the KAO. I'll close this one for now.

**Comment by cpanato** [comment]
thanks @kalmog

**Pull request closed without merging**